### ISSUE TITLE
Provide a way to share state across multiple compilations

### ISSUE THREAD

**Issue by znck** [sass/sass#2745]
When using sass libraries with react/vue components and building with webpack. Every imported file (in the component) is processed individually which requires the shared libraries to reload for every file. 

Ideally, JavaScript API can create ImportCache/AsyncImportCache by importing some shared files and libraries and pass this cache to every render/compile call.

**Comment by nex3** [comment]
This is something I brought up as a possibility as part of #2737. I'm retargeting it to more generally refer to a way of sharing state, because we'd need to expose this as part of the JS API that's shared between Dart Sass and Node Sass, and the latter doesn't have an identical notion of an `ImportCache`.

The reason we don't *automatically* share state is that there's always a possibility that, between different calls to `render()`/`renderSync()`, any given file has changed on the filesystem and so anything cached has become invalid. However, in a situation like webpack compiling many separate entrypoints that share a considerable amount of libraries, the client knows (or assumes) that the files won't change between those compilations.

Rather than exposing import caches to the JS API, I think it makes more sense to focus on expressing the guarantees that the user is providing the Sass implementation. For example, we could have a `Compiler` class that has `render()` and `renderSync()` as *methods*, and that explicitly assumes that the filesystem doesn't change during the lifetime of the `Compiler` instance.

We could even eventually extend this to provide the `Compiler` with notifications when files *do* change, so that it could invalidate only the cached data that's actually affected. This is essentially the model that we used in Ruby Sass as well.

**Comment by jquense** [comment]
I'd be interested in helping out with this, if there is a reasonably clear path forward on what the feature would look like. This issue is near and dear to me, all our projects use css-modules with sass and it results in hundreds of individually compiled sass files  with most referencing the same one or two "theme" files

**Comment by nex3** [comment]
Awesome! We haven't really had the opportunity to create a new shared JS API, so there's not a concrete process for it yet. Since it's a design that's shared across implementations, though, it probably makes sense to start with [a formal proposal](https://github.com/sass/sass/blob/master/CONTRIBUTING.md#proposal) in a similar style to a language proposal so all the stakeholders have a chance to chime in. Can you write something like that, but adapted for a JS API addition rather than a Sass language feature?

**Comment by jquense** [comment]
Yes! I can try and put something together, Sort of a design doc for the proposed API? No timeline on when i can get it done but adding it to my agenda :) Thanks for the help!

**Comment by nex3** [comment]
> Yes! I can try and put something together, Sort of a design doc for the proposed API?

That would be great :smile:

**Comment by illepic** [comment]
For more context to anyone searching around about this: not having a shared state within the Sass environment on a webpack/sass-loader project negates the value of Sass Modules and `@use`.

For example, a single conglomeration file like our [tokens here](https://github.com/phase2/particle/blob/841a6662119ee61ea37538733e4fd4bdd25eefb4/source/default/tokens/sass/index.scss) being referenced across a number of random components like [here ](https://github.com/phase2/particle/blob/841a6662119ee61ea37538733e4fd4bdd25eefb4/source/default/_patterns/01-atoms/svg/_svg.scss#L9)and [here](https://github.com/phase2/particle/blob/841a6662119ee61ea37538733e4fd4bdd25eefb4/source/default/_patterns/02-molecules/nav/_nav.scss#L2) just results in full CSS duplication due to webpack effectively "booting up" Sass on every reference.

**Comment by nex3** [comment]
I'm merging https://github.com/sass/sass/issues/2823 into this issue, and broadening the scope here beyond just compilations with unchanged sources. We can add a method to the `Compiler` that indicates when a file has changed, so that it can update its internal tree accordingly.
